File: AlggaGo1.0/physics.py
# physics.py
import pymunk
from pymunk import Vec2d
import random
import math 

# 설정
WIDTH, HEIGHT = 800, 800
MARGIN = 50
STONE_RADIUS = 25
STONE_MASS = 1

# 드래그 관련 상수
MAX_DRAG_LENGTH = 100    # (픽셀) 최대 허용 드래그 거리
FORCE_MULTIPLIER = 20     # (임펄스/픽셀) 드래그 길이당 곱해줄 배율
MIN_FORCE = 20          # (임펄스) 최소 기본 힘

def create_stone(space, position, color):
    moment = pymunk.moment_for_circle(STONE_MASS, 0, STONE_RADIUS)
    body = pymunk.Body(STONE_MASS, moment)
    body.position = position
    shape = pymunk.Circle(body, STONE_RADIUS)
    shape.collision_type = 1
    shape.elasticity = 1.0
    shape.friction = 0.9
    shape.color = (*color, 255)
    space.add(body, shape)
    return shape

def move_random_white(stones):
    white_stones = [s for s in stones if s.color[:3] == (255, 255, 255)]
    if not white_stones:
        return
    stone = random.choice(white_stones)

    angle = random.uniform(0, 2 * 3.14159)
    direction = Vec2d(1, 0).rotated(angle)
    impulse = direction * random.uniform(200, 500)
    stone.body.apply_impulse_at_world_point(impulse, stone.body.position)

# 첫 번째 reset_stones 함수 (주석 처리됨)
# def reset_stones(space, stones):
#     from physics import create_stone  # 재귀 방지용 (또는 top-level에서 import 하세요)
#     for shape in stones:
#         space.remove(shape, shape.body)
#     stones.clear()
# 
#     cell = (WIDTH - 2 * MARGIN) / 18
#     x_positions = [MARGIN + cell * i for i in range(3, 16, 4)]
#     y_top = MARGIN + STONE_RADIUS * 2 + 66.5
#     y_bottom = HEIGHT - MARGIN - STONE_RADIUS * 2 - 66.5
# 
#     for x in x_positions:
#         stones.append(create_stone(space, (x, y_top), color=(255, 255, 255)))
#     for x in x_positions:
#         stones.append(create_stone(space, (x, y_bottom), color=(0, 0, 0)))
# 
#     return 4, 4

def reset_stones(space, stones):
    """모드 1, 2, 3용 고정 배치 함수"""
    from physics import create_stone  # 재귀 방지용 (또는 top-level import)

    for shape in stones:
        space.remove(shape, shape.body)
    stones.clear()

    cell = (WIDTH - 2 * MARGIN) / 18
    x_positions = [MARGIN + cell * i for i in range(3, 16, 4)]
    y_top = MARGIN + STONE_RADIUS * 2 + 66.5
    y_bottom = HEIGHT - MARGIN - STONE_RADIUS * 2 - 66.5

    # --- 백돌 (고정 위치) ---
    for x in x_positions:
        stones.append(create_stone(space, (x, y_top), color=(255, 255, 255)))

    # --- 흑돌 (일직선 배치) ---
    # 하단에 4개 흑돌을 일직선으로 배치
    for x in x_positions:
        stones.append(create_stone(space, (x, y_bottom), color=(0, 0, 0)))

    return 4, 4  # 백돌, 흑돌 개수 반환


def reset_stones_random(space, stones):
    """모드 4용 랜덤 배치 함수"""
    from physics import create_stone  # 재귀 방지용 (또는 top-level import)
    import numpy as np

    for shape in stones:
        space.remove(shape, shape.body)
    stones.clear()

    cell = (WIDTH - 2 * MARGIN) / 18
    x_positions = [MARGIN + cell * i for i in range(3, 16, 4)]
    y_top = MARGIN + STONE_RADIUS * 2 + 66.5
    y_bottom = HEIGHT - MARGIN - STONE_RADIUS * 2 - 66.5

    # --- 백돌 (고정 위치) ---
    for x in x_positions:
        stones.append(create_stone(space, (x, y_top), color=(255, 255, 255)))

    # --- 흑돌 (충돌 없이 랜덤 배치) ---
    placed_positions = []
    max_attempts = 100

    def is_far_enough(x, y, placed, min_dist):
        for px, py in placed:
            if np.linalg.norm([x - px, y - py]) < min_dist:
                return False
        return True

    while len(placed_positions) < 4 and max_attempts > 0:
        x = np.random.uniform(MARGIN + STONE_RADIUS, WIDTH - MARGIN - STONE_RADIUS)
        y = y_bottom + np.random.uniform(-40, 40)

        if is_far_enough(x, y, placed_positions, STONE_RADIUS * 3.5):
            placed_positions.append((x, y))
            stones.append(create_stone(space, (x, y), color=(0, 0, 0)))
        else:
            max_attempts -= 1

    return 4, 4  # 백돌, 흑돌 개수 반환


def reset_stones_custom(space, stones, custom_black_positions):
    """커스텀 흑돌 위치로 돌을 배치하는 함수"""
    from physics import create_stone  # 재귀 방지용 (또는 top-level import)
    
    for shape in stones:
        space.remove(shape, shape.body)
    stones.clear()

    cell = (WIDTH - 2 * MARGIN) / 18
    x_positions = [MARGIN + cell * i for i in range(3, 16, 4)]
    y_top = MARGIN + STONE_RADIUS * 2 + 66.5

    # --- 백돌 (고정 위치) ---
    for x in x_positions:
        stones.append(create_stone(space, (x, y_top), color=(255, 255, 255)))

    # --- 흑돌 (커스텀 위치) ---
    for x, y in custom_black_positions:
        stones.append(create_stone(space, (x, y), color=(0, 0, 0)))

    return 4, 4  # 백돌, 흑돌 개수 반환


def all_stones_stopped(stones, threshold=5):
    """모든 돌의 속도가 충분히 느릴 때 True 반환"""
    for shape in stones:
        if shape.body.velocity.length > threshold:
            return False
    return True

def scale_force(force_normalized):
    """
    AI가 예측한 [0.0 ~ 1.0] force 값을 실제 물리 임펄스 값으로 변환
    """
    MAX_FORCE = MAX_DRAG_LENGTH * FORCE_MULTIPLIER
    return MIN_FORCE + force_normalized * (MAX_FORCE - MIN_FORCE)

# [추가] physics.py 파일에 아래 함수를 추가하세요.

def reset_stones_beginner(space, stones):
    """모드 6 (초보자용) 배치 함수 (백돌 4개 vs 흑돌 6개)"""
    from physics import create_stone

    for shape in stones:
        if shape in space.shapes:
            space.remove(shape, shape.body)
    stones.clear()

    cell = (WIDTH - 2 * MARGIN) / 18
    
    # --- 백돌 4개 (고정 위치) ---
    x_positions_white = [MARGIN + cell * i for i in range(3, 16, 4)]
    y_top = MARGIN + STONE_RADIUS * 2 + 66.5
    for x in x_positions_white:
        stones.append(create_stone(space, (x, y_top), color=(255, 255, 255)))

    # --- 흑돌 6개 (0.5칸 오른쪽 이동) ---
    # 18개 격자 중에서 중앙에 맞춰 넓게 분산하여 6개 위치 선택 (1.5, 4.5, 7.5, 10.5, 13.5, 16.5)
    x_positions_black = [MARGIN + cell * (i + 0.5) for i in [1, 4, 7, 10, 13, 16]]
    y_bottom = HEIGHT - MARGIN - STONE_RADIUS * 2 - 66.5
    for x in x_positions_black:
        stones.append(create_stone(space, (x, y_bottom), color=(0, 0, 0)))

    return 4, 6 # 백돌, 흑돌 개수 반환

File: docs/js/physics.js
// physics.js - Matter.js 기반 물리 엔진
const { Engine, Render, World, Bodies, Body, Vector, Composite } = Matter;

// 게임 설정 (Python physics.py와 동일)
const WIDTH = 800;
const HEIGHT = 800;
const MARGIN = 50;
const STONE_RADIUS = 25;
const STONE_MASS = 1;

// 드래그 관련 상수
const MAX_DRAG_LENGTH = 100;
const FORCE_MULTIPLIER = 20;
const MIN_FORCE = 20;

class PhysicsEngine {
    constructor() {
        this.engine = Engine.create();
        this.world = this.engine.world;
        
        // 중력 제거 (AlggaGo는 중력이 없는 게임)
        this.world.gravity.y = 0;
        
        // 경계 벽 생성
        this.createWalls();
        
        // 돌들을 저장할 배열
        this.stones = [];
        
        // 충돌 감지 설정
        this.setupCollisions();
    }
    
    createWalls() {
        const wallThickness = 10;
        const wallOptions = { 
            isStatic: true, 
            render: { fillStyle: '#34495e' } 
        };
        
        // 상단 벽
        this.topWall = Bodies.rectangle(WIDTH/2, MARGIN/2, WIDTH - 2*MARGIN, wallThickness, wallOptions);
        
        // 하단 벽
        this.bottomWall = Bodies.rectangle(WIDTH/2, HEIGHT - MARGIN/2, WIDTH - 2*MARGIN, wallThickness, wallOptions);
        
        // 좌측 벽
        this.leftWall = Bodies.rectangle(MARGIN/2, HEIGHT/2, wallThickness, HEIGHT - 2*MARGIN, wallOptions);
        
        // 우측 벽
        this.rightWall = Bodies.rectangle(WIDTH - MARGIN/2, HEIGHT/2, wallThickness, HEIGHT - 2*MARGIN, wallOptions);
        
        Composite.add(this.world, [this.topWall, this.bottomWall, this.leftWall, this.rightWall]);
    }
    
    setupCollisions() {
        // 충돌 이벤트 리스너
        Matter.Events.on(this.engine, 'collisionStart', (event) => {
            // 충돌 효과음 재생 (나중에 구현)
            console.log('Collision detected!');
        });
    }
    
    createStone(x, y, color) {
        const stone = Bodies.circle(x, y, STONE_RADIUS, {
            mass: STONE_MASS,
            restitution: 1.0, // 탄성
            friction: 0.9,
            render: {
                fillStyle: color === 'white' ? '#ffffff' : '#000000',
                strokeStyle: '#333333',
                lineWidth: 2
            }
        });
        
        // 돌에 메타데이터 추가
        stone.color = color;
        stone.isStone = true;
        
        Composite.add(this.world, stone);
        this.stones.push(stone);
        
        return stone;
    }
    
    resetStones() {
        // 기존 돌들 제거
        this.stones.forEach(stone => {
            Composite.remove(this.world, stone);
        });
        this.stones = [];
        
        // 격자 계산
        const cell = (WIDTH - 2 * MARGIN) / 18;
        const xPositions = [];
        for (let i = 3; i <= 15; i += 4) {
            xPositions.push(MARGIN + cell * i);
        }
        
        const yTop = MARGIN + STONE_RADIUS * 2 + 66.5;
        const yBottom = HEIGHT - MARGIN - STONE_RADIUS * 2 - 66.5;
        
        // 백돌 (상단)
        xPositions.forEach(x => {
            this.createStone(x, yTop, 'white');
        });
        
        // 흑돌 (하단)
        xPositions.forEach(x => {
            this.createStone(x, yBottom, 'black');
        });
        
        return { white: 4, black: 4 };
    }
    
    applyImpulse(stone, impulse) {
        Body.applyForce(stone, stone.position, impulse);
    }
    
    allStonesStopped() {
        const velocityThreshold = 0.5;
        return this.stones.every(stone => {
            const velocity = stone.velocity;
            return Math.abs(velocity.x) < velocityThreshold && Math.abs(velocity.y) < velocityThreshold;
        });
    }
    
    getStonesByColor(color) {
        return this.stones.filter(stone => stone.color === color);
    }
    
    removeStone(stone) {
        Composite.remove(this.world, stone);
        const index = this.stones.indexOf(stone);
        if (index > -1) {
            this.stones.splice(index, 1);
        }
    }
    
    update() {
        Engine.update(this.engine, 1000 / 60); // 60fps
        
        // 화면 밖으로 나간 돌들 제거
        this.stones = this.stones.filter(stone => {
            const pos = stone.position;
            if (pos.x < 0 || pos.x > WIDTH || pos.y < 0 || pos.y > HEIGHT) {
                Composite.remove(this.world, stone);
                return false;
            }
            return true;
        });
    }
    
    // 드래그 힘 계산
    calculateImpulse(dragStart, dragEnd) {
        const dragVector = Vector.sub(dragStart, dragEnd);
        const length = Vector.magnitude(dragVector);
        
        if (length === 0) return Vector.create(0, 0);
        
        const clampedLength = Math.min(length, MAX_DRAG_LENGTH);
        const normalizedVector = Vector.normalise(dragVector);
        const force = (clampedLength * FORCE_MULTIPLIER + MIN_FORCE) / 1000; // 스케일 조정
        
        return Vector.mult(normalizedVector, force);
    }
}

// 전역 변수로 export
window.PhysicsEngine = PhysicsEngine;
window.GAME_CONSTANTS = {
    WIDTH, HEIGHT, MARGIN, STONE_RADIUS, STONE_MASS,
    MAX_DRAG_LENGTH, FORCE_MULTIPLIER, MIN_FORCE
};
